# Capstone Project on Data Cleansing

### Importing Necessary Modules for performing Data cleansing in Python: -

NumPy: is a library for the Python programming language, adding support for large, multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions to operate on these arrays.

Pandas: is a software library written for the Python programming language for data manipulation and analysis. In particular, it offers data structures and operations for manipulating numerical tables and time series.

Matplotlib: is a plotting library for the Python programming language and its numerical mathematics extension NumPy. It provides an object-oriented API for embedding plots into applications using general-purpose GUI toolkits like Tkinter, wxPython, Qt, or GTK.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Read the Melbourne dataset:

In [2]:
df = pd.read_csv(r'Downloads\Melbourne.csv')

### Show the dataset imported with all fields:

In [3]:
df.head(5)

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  03-09-2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  03-12-2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  04-02-2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  04-02-2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  04-03-2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

### Know your Data: -

In [4]:
df.shape

(23547, 21)

### First Check how many fields contain NAN(Not A Number) values: -
** Very Important step **

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23547 entries, 0 to 23546
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         23547 non-null  object 
 1   Address        23547 non-null  object 
 2   Rooms          23547 non-null  int64  
 3   Type           23547 non-null  object 
 4   Price          18396 non-null  float64
 5   Method         23547 non-null  object 
 6   SellerG        23547 non-null  object 
 7   Date           23547 non-null  object 
 8   Distance       23546 non-null  float64
 9   Postcode       23546 non-null  float64
 10  Bedroom2       19066 non-null  float64
 11  Bathroom       19063 non-null  float64
 12  Car            18921 non-null  float64
 13  Landsize       17410 non-null  float64
 14  BuildingArea   10018 non-null  float64
 15  YearBuilt      11540 non-null  float64
 16  CouncilArea    15656 non-null  object 
 17  Lattitude      19243 non-null  float64
 18  Longti

### As in the above step we can see each column contains 23547 entries but some columns have missing entries like Price column has 18396 entries rest are NAN or missing and so on with Bedroom2, Bathroom, Car, etc....

### Lets sum up all the missing value entries column-wise before presenting reports: -

In [6]:
# will show TRUE wherever Null entries in all the columns

df.isnull()

Suburb  Address  Rooms   Type  Price  Method  SellerG   Date  Distance  \
0       False    False  False  False   True   False    False  False     False   
1       False    False  False  False  False   False    False  False     False   
2       False    False  False  False  False   False    False  False     False   
3       False    False  False  False   True   False    False  False     False   
4       False    False  False  False  False   False    False  False     False   
...       ...      ...    ...    ...    ...     ...      ...    ...       ...   
23542   False    False  False  False   True   False    False  False     False   
23543   False    False  False  False   True   False    False  False     False   
23544   False    False  False  False  False   False    False  False     False   
23545   False    False  False  False  False   False    False  False     False   
23546   False    False  False  False  False   False    False  False     False   

       Postcode  ...  Bathroom    Car  Landsize  BuildingArea  YearBuilt  \
0         False  ...     False  False     False          True       True   
1         False  ...     False  False     False          True       True   
2         False  ...     False  False     False         False      False   
3         False  ...     False  False     False          True       True   
4         False  ...     False  False     False         False      False   
...         ...  ...       ...    ...       ...           ...        ...   
23542     False  ...     False  False     False         False      False   
23543     False  ...      True   True      True          True       True   
23544     False  ...     False  False      True          True       True   
23545     False  ...     False  False     False         False      False   
23546     False  ...     False  False      True         False      False   

       CouncilArea  Lattitude  Longtitude  Regionname  Propertycount  
0            False      False       False       False          False  
1            False      False       False       False          False  
2            False      False       False       False          False  
3            False      False       False       False          False  
4            False      False       False       False          False  
...            ...        ...         ...         ...            ...  
23542         True      False       False       False          False  
23543         True      False       False       False          False  
23544         True      False       False       False          False  
23545         True      False       False       False          False  
23546         True      False       False       False          False  

[23547 rows x 21 columns]

In [8]:
# we can summarize the columns having Null entries.

col = df.isnull().any()  # or we can use df.isnull() > 0
col

Suburb           False
Address          False
Rooms            False
Type             False
Price             True
Method           False
SellerG          False
Date             False
Distance          True
Postcode          True
Bedroom2          True
Bathroom          True
Car               True
Landsize          True
BuildingArea      True
YearBuilt         True
CouncilArea       True
Lattitude         True
Longtitude        True
Regionname        True
Propertycount     True
dtype: bool

In [7]:
# Now we sum all the TRUE entries in each column (by default column-wise)

df.isnull().sum()  # sum(axis = 0)

Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

In [9]:
# to check all the columns or Indexes column-wise

col.index  # or col.index[True]

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [10]:
# to check the index values column-wise

col.values

array([False, False, False, False,  True, False, False, False,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True])

In [11]:
# prints the Index value at 0th index

col.index[0]

'Suburb'

In [12]:
# prints all the columns with Null values = True

col_null = col.index[col.values]
col_null

Index(['Price', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [13]:
# We can sort the columns based on their Null values (by default in Ascending)

df.isnull().sum(axis = 0).sort_values()  # or sort_values(ascending = True)

Suburb               0
Address              0
Rooms                0
Type                 0
Method               0
SellerG              0
Date                 0
Regionname           1
Postcode             1
Propertycount        1
Distance             1
Lattitude         4304
Longtitude        4304
Bedroom2          4481
Bathroom          4484
Car               4626
Price             5151
Landsize          6137
CouncilArea       7891
YearBuilt        12007
BuildingArea     13529
dtype: int64

In [14]:
# To sort Columns in Descending order based on their missing values

df.isnull().sum(axis = 0).sort_values(ascending = False)

BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Landsize          6137
Price             5151
Car               4626
Bathroom          4484
Bedroom2          4481
Longtitude        4304
Lattitude         4304
Distance             1
Propertycount        1
Postcode             1
Regionname           1
Date                 0
SellerG              0
Method               0
Type                 0
Rooms                0
Address              0
Suburb               0
dtype: int64

In [15]:
# Before moving ahead we can also check if there is any column having all Null values

df.isnull().all()  # all() by default column-wise

Suburb           False
Address          False
Rooms            False
Type             False
Price            False
Method           False
SellerG          False
Date             False
Distance         False
Postcode         False
Bedroom2         False
Bathroom         False
Car              False
Landsize         False
BuildingArea     False
YearBuilt        False
CouncilArea      False
Lattitude        False
Longtitude       False
Regionname       False
Propertycount    False
dtype: bool

### We can also check missing values Row-wise (Not used very often)

In [16]:
df.isnull().sum(axis = 1)  

0        3
1        2
2        0
3        3
4        0
        ..
23542    2
23543    8
23544    4
23545    1
23546    2
Length: 23547, dtype: int64

In [17]:
# to check the BOOLEAN values of each row with Null entries

row = df.isnull().any(axis = 1)
row

0         True
1         True
2        False
3         True
4        False
         ...  
23542     True
23543     True
23544     True
23545     True
23546     True
Length: 23547, dtype: bool

In [18]:
# to check the index values row-wise

row.index

RangeIndex(start=0, stop=23547, step=1)

In [19]:
# to check values associated with each index row-wise

row.values

array([ True,  True, False, ...,  True,  True,  True])

In [20]:
# prints all the rows with Null values = True

row_null = row.index[row.values]
row_null

Int64Index([    0,     1,     3,     5,     7,     8,     9,    10,    12,
               13,
            ...
            23537, 23538, 23539, 23540, 23541, 23542, 23543, 23544, 23545,
            23546],
           dtype='int64', length=17351)

In [21]:
df.isnull().sum(axis = 1).sort_values()

8551      0
11550     0
11551     0
5077      0
11553     0
         ..
6313     10
20968    10
16232    10
17100    10
18523    13
Length: 23547, dtype: int64

In [22]:
# we can also print all the dataset now, with atleast one Null value in each row
# (ie; if a row does not consists Null value it will not be printed).

df.iloc[row_null,:]  # iloc[table, no of rows to print]

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St      2    h        NaN     SS   
1        Abbotsford        85 Turner St      2    h  1480000.0      S   
3        Abbotsford  18/659 Victoria St      3    u        NaN     VB   
5        Abbotsford    40 Federation La      3    h   850000.0     PI   
7        Abbotsford        16 Maugie St      4    h        NaN     SN   
...             ...                 ...    ...  ...        ...    ...   
23542  Wyndham Vale     25 Clitheroe Dr      3    u        NaN     PN   
23543  Wyndham Vale    19 Dalrymple Bvd      4    h        NaN      S   
23544     Yallambie        17 Amaroo Wy      4    h  1100000.0      S   
23545    Yarraville          6 Agnes St      4    h  1285000.0     SP   
23546    Yarraville       33 Freeman St      4    h  1050000.0     VB   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0  ...       1.0  1.0   
1             Biggin  03-12-2016       2.5    3067.0  ...       1.0  1.0   
3             Rounds  04-02-2016       2.5    3067.0  ...       2.0  1.0   
5             Biggin  04-03-2017       2.5    3067.0  ...       2.0  1.0   
7             Nelson  06-08-2016       2.5    3067.0  ...       2.0  2.0   
...              ...         ...       ...       ...  ...       ...  ...   
23542      Harcourts  26-08-2017      27.2    3024.0  ...       1.0  0.0   
23543  hockingstuart  26-08-2017      27.2    3024.0  ...       NaN  NaN   
23544     Buckingham  26-08-2017      12.7    3085.0  ...       3.0  2.0   
23545        Village  26-08-2017       6.3    3013.0  ...       1.0  1.0   
23546        Village  26-08-2017       6.3    3013.0  ...       2.0  2.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0         126.0           NaN        NaN        Yarra -37.80140   144.99580   
1         202.0           NaN        NaN        Yarra -37.79960   144.99840   
3           0.0           NaN        NaN        Yarra -37.81140   145.01160   
5          94.0           NaN        NaN        Yarra -37.79690   144.99690   
7         400.0         220.0     2006.0        Yarra -37.79650   144.99650   
...         ...           ...        ...          ...       ...         ...   
23542     552.0         119.0     1990.0          NaN -37.90032   144.61839   
23543       NaN           NaN        NaN          NaN -37.87882   144.60184   
23544       NaN           NaN        NaN          NaN -37.72006   145.10547   
23545     362.0         112.0     1920.0          NaN -37.81188   144.88449   
23546       NaN         139.0     1950.0          NaN -37.81829   144.87404   

                  Regionname Propertycount  
0      Northern Metropolitan        4019.0  
1      Northern Metropolitan        4019.0  
3      Northern Metropolitan        4019.0  
5      Northern Metropolitan        4019.0  
7      Northern Metropolitan        4019.0  
...                      ...           ...  
23542   Western Metropolitan        5262.0  
23543   Western Metropolitan        5262.0  
23544  Northern Metropolitan        1369.0  
23545   Western Metropolitan        6543.0  
23546   Western Metropolitan        6543.0  

[17351 rows x 21 columns]

In [23]:
# to check if any row consists all Null values only

df.isnull().all(axis=1)  # or df.isnull().sum(axis = 1) == 21 (since each row consists 21 entries)

0        False
1        False
2        False
3        False
4        False
         ...  
23542    False
23543    False
23544    False
23545    False
23546    False
Length: 23547, dtype: bool

In [24]:
# since there are large number of rows, we can summarize

df.isnull().all(axis=1).sum()

0

### We can also summarise the Null values %wise

In [25]:
prcnt1 = df.isnull().sum().sort_values()/len(df) * 100
prcnt1

Suburb            0.000000
Address           0.000000
Rooms             0.000000
Type              0.000000
Method            0.000000
SellerG           0.000000
Date              0.000000
Regionname        0.004247
Postcode          0.004247
Propertycount     0.004247
Distance          0.004247
Lattitude        18.278337
Longtitude       18.278337
Bedroom2         19.030025
Bathroom         19.042766
Car              19.645815
Price            21.875398
Landsize         26.062768
CouncilArea      33.511700
YearBuilt        50.991634
BuildingArea     57.455302
dtype: float64

In [26]:
# we can sort according to %ge of missing values

prcnt2 = df.isnull().sum().sort_values(ascending = False)/len(df) * 100
prcnt2

BuildingArea     57.455302
YearBuilt        50.991634
CouncilArea      33.511700
Landsize         26.062768
Price            21.875398
Car              19.645815
Bathroom         19.042766
Bedroom2         19.030025
Longtitude       18.278337
Lattitude        18.278337
Distance          0.004247
Propertycount     0.004247
Postcode          0.004247
Regionname        0.004247
Date              0.000000
SellerG           0.000000
Method            0.000000
Type              0.000000
Rooms             0.000000
Address           0.000000
Suburb            0.000000
dtype: float64

In [27]:
# lets round of decimals to 2 place for clear view of missing values

round(prcnt2, 2)  # or round(df.isnull().sum().sort_values(ascending = False)/len(df) * 100, 2) is much faster

BuildingArea     57.46
YearBuilt        50.99
CouncilArea      33.51
Landsize         26.06
Price            21.88
Car              19.65
Bathroom         19.04
Bedroom2         19.03
Longtitude       18.28
Lattitude        18.28
Distance          0.00
Propertycount     0.00
Postcode          0.00
Regionname        0.00
Date              0.00
SellerG           0.00
Method            0.00
Type              0.00
Rooms             0.00
Address           0.00
Suburb            0.00
dtype: float64

In [28]:
# If we want check %ge of Null values Row-wise

round(df.isnull().sum(axis = 1).sort_values(ascending = False)/len(df) * 100, 2)

18523    0.06
17100    0.04
16232    0.04
20968    0.04
6313     0.04
         ... 
11553    0.00
5077     0.00
11551    0.00
11550    0.00
8551     0.00
Length: 23547, dtype: float64

### Now we can see the top 3 columns with very high %ge of Missing/Null values, which is not necessary for our Data Analysis or making any Interpretation, so we can drop these columns

In [29]:
noise = prcnt2.head(3).index.values
noise

array(['BuildingArea', 'YearBuilt', 'CouncilArea'], dtype=object)

In [30]:
df = df.drop(noise, axis = 'columns')
df

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St      2    h        NaN     SS   
1        Abbotsford        85 Turner St      2    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St      2    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St      3    u        NaN     VB   
4        Abbotsford        5 Charles St      3    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23542  Wyndham Vale     25 Clitheroe Dr      3    u        NaN     PN   
23543  Wyndham Vale    19 Dalrymple Bvd      4    h        NaN      S   
23544     Yallambie        17 Amaroo Wy      4    h  1100000.0      S   
23545    Yarraville          6 Agnes St      4    h  1285000.0     SP   
23546    Yarraville       33 Freeman St      4    h  1050000.0     VB   

             SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0       2.0       1.0  1.0   
1             Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2             Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
3             Rounds  04-02-2016       2.5    3067.0       3.0       2.0  1.0   
4             Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
...              ...         ...       ...       ...       ...       ...  ...   
23542      Harcourts  26-08-2017      27.2    3024.0       3.0       1.0  0.0   
23543  hockingstuart  26-08-2017      27.2    3024.0       NaN       NaN  NaN   
23544     Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23545        Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23546        Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
0         126.0  -37.80140   144.99580  Northern Metropolitan         4019.0  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
3           0.0  -37.81140   145.01160  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23542     552.0  -37.90032   144.61839   Western Metropolitan         5262.0  
23543       NaN  -37.87882   144.60184   Western Metropolitan         5262.0  
23544       NaN  -37.72006   145.10547  Northern Metropolitan         1369.0  
23545     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23546       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[23547 rows x 18 columns]

In [31]:
round(df.isnull().sum().sort_values(ascending = False)/len(df) * 100, 2)

Landsize         26.06
Price            21.88
Car              19.65
Bathroom         19.04
Bedroom2         19.03
Longtitude       18.28
Lattitude        18.28
Propertycount     0.00
Regionname        0.00
Distance          0.00
Postcode          0.00
Date              0.00
SellerG           0.00
Method            0.00
Type              0.00
Rooms             0.00
Address           0.00
Suburb            0.00
dtype: float64

### Now we can also check the rows having more than 5 Null values which might also not be necessary for our Analysis purposes- 

In [33]:
df[df.isnull().sum(axis = 1) > 5]  # here we can see out of total 23547 rows 4278 rows has more than 5 null values

Suburb              Address  Rooms Type      Price Method  \
15            Abbotsford     217 Langridge St      3    h  1000000.0      S   
16            Abbotsford      18a Mollison St      2    t   745000.0      S   
19            Abbotsford  403/609 Victoria St      2    u   542000.0      S   
21            Abbotsford    25/84 Trenerry Cr      2    u   760000.0     SP   
22            Abbotsford    106/119 Turner St      1    u   481000.0     SP   
...                  ...                  ...    ...  ...        ...    ...   
23495  Templestowe Lower  1/207 Manningham Rd      2    u   550000.0     PI   
23499          Thornbury      1/128 Dundas St      2    t   770000.0     PI   
23502          Thornbury        111 Pender St      2    t   858000.0      S   
23508             Toorak   21/1059 Malvern Rd      2    u   720000.0     VB   
23531           Werribee         91 Latham St      4    h   540000.0     PI   

            SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
15           Jellis  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
16           Jellis  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
19           Dingle  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
21           Biggin  10-12-2016       2.5    3067.0       NaN       NaN  NaN   
22     Purplebricks  10-12-2016       2.5    3067.0       NaN       NaN  NaN   
...             ...         ...       ...       ...       ...       ...  ...   
23495         Barry  26-08-2017      12.4    3107.0       NaN       NaN  NaN   
23499       McGrath  26-08-2017       7.0    3071.0       NaN       NaN  NaN   
23502        Jellis  26-08-2017       7.0    3071.0       NaN       NaN  NaN   
23508        Beller  26-08-2017       4.1    3142.0       NaN       NaN  NaN   
23531       Triwest  26-08-2017      14.7    3030.0       NaN       NaN  NaN   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
15          NaN        NaN         NaN  Northern Metropolitan         4019.0  
16          NaN        NaN         NaN  Northern Metropolitan         4019.0  
19          NaN        NaN         NaN  Northern Metropolitan         4019.0  
21          NaN        NaN         NaN  Northern Metropolitan         4019.0  
22          NaN        NaN         NaN  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23495       NaN        NaN         NaN   Eastern Metropolitan         5420.0  
23499       NaN        NaN         NaN  Northern Metropolitan         8870.0  
23502       NaN        NaN         NaN  Northern Metropolitan         8870.0  
23508       NaN        NaN         NaN  Southern Metropolitan         7217.0  
23531       NaN        NaN         NaN   Western Metropolitan        16166.0  

[4278 rows x 18 columns]

In [34]:
# we can also calculate total %ge of rows with more than 5 Null values in our whole data by using:

len(df[df.isnull().sum(axis = 1) > 5])/len(df) * 100  # ie: 18.17% of rows contains more than 5 Null values

18.16791948018856

In [37]:
# Now lets retain the rows having less than 5 Null values

df.isnull().sum(axis = 1) <= 5  # here we can see 19269 rows contains less than 5 Null values

0        True
1        True
2        True
3        True
4        True
         ... 
23542    True
23543    True
23544    True
23545    True
23546    True
Length: 19269, dtype: bool

### Now we can update our dataset 'df' finally with the Null values less than 5 Null values

In [36]:
df = df[df.isnull().sum(axis = 1) <= 5]
df

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St      2    h        NaN     SS   
1        Abbotsford        85 Turner St      2    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St      2    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St      3    u        NaN     VB   
4        Abbotsford        5 Charles St      3    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23542  Wyndham Vale     25 Clitheroe Dr      3    u        NaN     PN   
23543  Wyndham Vale    19 Dalrymple Bvd      4    h        NaN      S   
23544     Yallambie        17 Amaroo Wy      4    h  1100000.0      S   
23545    Yarraville          6 Agnes St      4    h  1285000.0     SP   
23546    Yarraville       33 Freeman St      4    h  1050000.0     VB   

             SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0       2.0       1.0  1.0   
1             Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2             Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
3             Rounds  04-02-2016       2.5    3067.0       3.0       2.0  1.0   
4             Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
...              ...         ...       ...       ...       ...       ...  ...   
23542      Harcourts  26-08-2017      27.2    3024.0       3.0       1.0  0.0   
23543  hockingstuart  26-08-2017      27.2    3024.0       NaN       NaN  NaN   
23544     Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23545        Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23546        Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
0         126.0  -37.80140   144.99580  Northern Metropolitan         4019.0  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
3           0.0  -37.81140   145.01160  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23542     552.0  -37.90032   144.61839   Western Metropolitan         5262.0  
23543       NaN  -37.87882   144.60184   Western Metropolitan         5262.0  
23544       NaN  -37.72006   145.10547  Northern Metropolitan         1369.0  
23545     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23546       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[19269 rows x 18 columns]

In [39]:
round(df.isnull().sum().sort_values(ascending = False)/len(df) * 100, 2)

Price            21.71
Landsize          9.65
Car               1.81
Bathroom          1.07
Bedroom2          1.05
Longtitude        0.13
Lattitude         0.13
Propertycount     0.00
Method            0.00
Address           0.00
Rooms             0.00
Type              0.00
Distance          0.00
SellerG           0.00
Date              0.00
Regionname        0.00
Postcode          0.00
Suburb            0.00
dtype: float64

### Now we can see that Price column still consists of more %ge(21.71%) of Null values but Price is very important column for Interpretaion hence we cannot drop the whole Price column, but here we can remove the rows which has Price values = Null or NaN

### To do this we Have 2 methods in Python

### Method 1: -

In [48]:
# lets first describe the Price column here we can see that we cannot 
# Impute the Null values of Price with other values of the column, hence we remove:


df['Price'].describe()

count    1.508600e+04
mean     1.066695e+06
std      6.389331e+05
min      8.500000e+04
25%      6.400000e+05
50%      9.000000e+05
75%      1.320000e+06
max      9.000000e+06
Name: Price, dtype: float64

In [40]:
df['Price'].notnull()

0        False
1         True
2         True
3        False
4         True
         ...  
23542    False
23543    False
23544     True
23545     True
23546     True
Name: Price, Length: 19269, dtype: bool

In [41]:
# here lets pass this as Index value to our dataset 'df', but let's not update df yet:

df[df['Price'].notnull()]

Suburb           Address  Rooms Type      Price Method  \
1        Abbotsford      85 Turner St      2    h  1480000.0      S   
2        Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
4        Abbotsford      5 Charles St      3    h  1465000.0     SP   
5        Abbotsford  40 Federation La      3    h   850000.0     PI   
6        Abbotsford       55a Park St      4    h  1600000.0     VB   
...             ...               ...    ...  ...        ...    ...   
23540  Williamstown   8/2 Thompson St      2    t   622500.0     SP   
23541  Williamstown      96 Verdon St      4    h  2500000.0     PI   
23544     Yallambie      17 Amaroo Wy      4    h  1100000.0      S   
23545    Yarraville        6 Agnes St      4    h  1285000.0     SP   
23546    Yarraville     33 Freeman St      4    h  1050000.0     VB   

          SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
1          Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2          Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
4          Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
5          Biggin  04-03-2017       2.5    3067.0       3.0       2.0  1.0   
6          Nelson  04-06-2016       2.5    3067.0       3.0       1.0  2.0   
...           ...         ...       ...       ...       ...       ...  ...   
23540        Greg  26-08-2017       6.8    3016.0       2.0       2.0  1.0   
23541     Sweeney  26-08-2017       6.8    3016.0       4.0       1.0  5.0   
23544  Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23545     Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23546     Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
5          94.0  -37.79690   144.99690  Northern Metropolitan         4019.0  
6         120.0  -37.80720   144.99410  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23540       NaN  -37.86393   144.90484   Western Metropolitan         6380.0  
23541     866.0  -37.85908   144.89299   Western Metropolitan         6380.0  
23544       NaN  -37.72006   145.10547  Northern Metropolitan         1369.0  
23545     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23546       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[15086 rows x 18 columns]

### Method 2: -

In [43]:
[~np.isnan(df['Price'])]

[0        False
 1         True
 2         True
 3        False
 4         True
          ...  
 23542    False
 23543    False
 23544     True
 23545     True
 23546     True
 Name: Price, Length: 19269, dtype: bool]

In [44]:
# here lets pass this as Index value to our dataset 'df', but let's not update df yet:

df[~np.isnan(df['Price'])]

Suburb           Address  Rooms Type      Price Method  \
1        Abbotsford      85 Turner St      2    h  1480000.0      S   
2        Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
4        Abbotsford      5 Charles St      3    h  1465000.0     SP   
5        Abbotsford  40 Federation La      3    h   850000.0     PI   
6        Abbotsford       55a Park St      4    h  1600000.0     VB   
...             ...               ...    ...  ...        ...    ...   
23540  Williamstown   8/2 Thompson St      2    t   622500.0     SP   
23541  Williamstown      96 Verdon St      4    h  2500000.0     PI   
23544     Yallambie      17 Amaroo Wy      4    h  1100000.0      S   
23545    Yarraville        6 Agnes St      4    h  1285000.0     SP   
23546    Yarraville     33 Freeman St      4    h  1050000.0     VB   

          SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
1          Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2          Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
4          Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
5          Biggin  04-03-2017       2.5    3067.0       3.0       2.0  1.0   
6          Nelson  04-06-2016       2.5    3067.0       3.0       1.0  2.0   
...           ...         ...       ...       ...       ...       ...  ...   
23540        Greg  26-08-2017       6.8    3016.0       2.0       2.0  1.0   
23541     Sweeney  26-08-2017       6.8    3016.0       4.0       1.0  5.0   
23544  Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23545     Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23546     Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
5          94.0  -37.79690   144.99690  Northern Metropolitan         4019.0  
6         120.0  -37.80720   144.99410  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23540       NaN  -37.86393   144.90484   Western Metropolitan         6380.0  
23541     866.0  -37.85908   144.89299   Western Metropolitan         6380.0  
23544       NaN  -37.72006   145.10547  Northern Metropolitan         1369.0  
23545     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23546       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[15086 rows x 18 columns]

### Now using any method 1 or 2 we can update our dataset 'df' finally

In [46]:
df = df[df['Price'].notnull()]
df

Suburb           Address  Rooms Type      Price Method  \
1        Abbotsford      85 Turner St      2    h  1480000.0      S   
2        Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
4        Abbotsford      5 Charles St      3    h  1465000.0     SP   
5        Abbotsford  40 Federation La      3    h   850000.0     PI   
6        Abbotsford       55a Park St      4    h  1600000.0     VB   
...             ...               ...    ...  ...        ...    ...   
23540  Williamstown   8/2 Thompson St      2    t   622500.0     SP   
23541  Williamstown      96 Verdon St      4    h  2500000.0     PI   
23544     Yallambie      17 Amaroo Wy      4    h  1100000.0      S   
23545    Yarraville        6 Agnes St      4    h  1285000.0     SP   
23546    Yarraville     33 Freeman St      4    h  1050000.0     VB   

          SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
1          Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2          Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
4          Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
5          Biggin  04-03-2017       2.5    3067.0       3.0       2.0  1.0   
6          Nelson  04-06-2016       2.5    3067.0       3.0       1.0  2.0   
...           ...         ...       ...       ...       ...       ...  ...   
23540        Greg  26-08-2017       6.8    3016.0       2.0       2.0  1.0   
23541     Sweeney  26-08-2017       6.8    3016.0       4.0       1.0  5.0   
23544  Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23545     Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23546     Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
5          94.0  -37.79690   144.99690  Northern Metropolitan         4019.0  
6         120.0  -37.80720   144.99410  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23540       NaN  -37.86393   144.90484   Western Metropolitan         6380.0  
23541     866.0  -37.85908   144.89299   Western Metropolitan         6380.0  
23544       NaN  -37.72006   145.10547  Northern Metropolitan         1369.0  
23545     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23546       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[15086 rows x 18 columns]

In [47]:
# here we can see the dataset having no Null values in Price column

round(df.isnull().sum().sort_values(ascending = False)/len(df) * 100, 2)

Landsize         9.83
Car              1.76
Bathroom         1.07
Bedroom2         1.05
Longtitude       0.15
Lattitude        0.15
Propertycount    0.00
Method           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Distance         0.00
SellerG          0.00
Date             0.00
Regionname       0.00
Postcode         0.00
Suburb           0.00
dtype: float64

### Here Landsize also has little high Null Values but since it is also an important column for us so just like Price column we can drop the rows containing Null values in Landsize

In [50]:
# lets first describe the Landsize column where we can see that we cannot 
# Impute the Null values of Landsize with other values of the column, hence we remove:

df['Landsize'].describe()

count     13603.000000
mean        558.116371
std        3987.326586
min           0.000000
25%         176.500000
50%         440.000000
75%         651.000000
max      433014.000000
Name: Landsize, dtype: float64

In [53]:
df = df[df['Landsize'].notnull()]
df

Suburb           Address  Rooms Type      Price Method  \
1         Abbotsford      85 Turner St      2    h  1480000.0      S   
2         Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
4         Abbotsford      5 Charles St      3    h  1465000.0     SP   
5         Abbotsford  40 Federation La      3    h   850000.0     PI   
6         Abbotsford       55a Park St      4    h  1600000.0     VB   
...              ...               ...    ...  ...        ...    ...   
23537  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
23538   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
23539   Williamstown       83 Power St      3    h  1170000.0      S   
23541   Williamstown      96 Verdon St      4    h  2500000.0     PI   
23545     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
1        Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2        Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
4        Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
5        Biggin  04-03-2017       2.5    3067.0       3.0       2.0  1.0   
6        Nelson  04-06-2016       2.5    3067.0       3.0       1.0  2.0   
...         ...         ...       ...       ...       ...       ...  ...   
23537     Barry  26-08-2017      16.7    3150.0       4.0       2.0  2.0   
23538  Williams  26-08-2017       6.8    3016.0       3.0       2.0  2.0   
23539     Raine  26-08-2017       6.8    3016.0       3.0       2.0  4.0   
23541   Sweeney  26-08-2017       6.8    3016.0       4.0       1.0  5.0   
23545   Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   

       Landsize  Lattitude  Longtitude                  Regionname  \
1         202.0  -37.79960   144.99840       Northern Metropolitan   
2         156.0  -37.80790   144.99340       Northern Metropolitan   
4         134.0  -37.80930   144.99440       Northern Metropolitan   
5          94.0  -37.79690   144.99690       Northern Metropolitan   
6         120.0  -37.80720   144.99410       Northern Metropolitan   
...         ...        ...         ...                         ...   
23537     652.0  -37.90562   145.16761  South-Eastern Metropolitan   
23538     333.0  -37.85927   144.87904        Western Metropolitan   
23539     436.0  -37.85274   144.88738        Western Metropolitan   
23541     866.0  -37.85908   144.89299        Western Metropolitan   
23545     362.0  -37.81188   144.88449        Western Metropolitan   

       Propertycount  
1             4019.0  
2             4019.0  
4             4019.0  
5             4019.0  
6             4019.0  
...              ...  
23537         7392.0  
23538         6380.0  
23539         6380.0  
23541         6380.0  
23545         6543.0  

[13603 rows x 18 columns]

In [54]:
round(df.isnull().sum().sort_values(ascending = False)/len(df) * 100, 2)

Car              0.46
Longtitude       0.16
Lattitude        0.16
Bathroom         0.01
Propertycount    0.00
SellerG          0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
Distance         0.00
Date             0.00
Regionname       0.00
Postcode         0.00
Bedroom2         0.00
Landsize         0.00
Suburb           0.00
dtype: float64

### As we can see Lattitude and Longitude has very few Null values, so rather dropping the rows we can Impute the Null values with Mean of the values of individual attributes.

In [55]:
df.loc[:, ['Lattitude', 'Longtitude']].describe()  # here we can describe the mean values first before imputing

Lattitude    Longtitude
count  13581.000000  13581.000000
mean     -37.809204    144.995221
std        0.079257      0.103913
min      -38.182550    144.431810
25%      -37.856820    144.929600
50%      -37.802360    145.000100
75%      -37.756400    145.058320
max      -37.408530    145.526350

In [56]:
df['Lattitude'].mean()

-37.80920350636918

In [58]:
df.loc[:, 'Lattitude'].fillna(df['Lattitude'].mean(), inplace = True)  
# we can use inplace fn to update 'df' instead updating and the assignining it using 
# df = df.loc[:, 'Lattitude'].fillna(df['Lattitude'].mean()), it means same

In [57]:
df['Longtitude'].mean()

144.995220860761

In [59]:
df.loc[:, 'Longtitude'].fillna(df['Longtitude'].mean(), inplace = True)

In [60]:
round(df.isnull().sum().sort_values(ascending = False)/len(df) * 100, 2)

Car              0.46
Bathroom         0.01
Propertycount    0.00
Date             0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Distance         0.00
Regionname       0.00
Postcode         0.00
Bedroom2         0.00
Landsize         0.00
Lattitude        0.00
Longtitude       0.00
Suburb           0.00
dtype: float64

### Now let us describe Car and Bathroom columns and clean their data also

In [62]:
df.loc[:, ['Car', 'Bathroom']].describe()  # here we can describe the mean values first before imputing

Car      Bathroom
count  13540.000000  13602.000000
mean       1.610414      1.534921
std        0.962244      0.691834
min        0.000000      0.000000
25%        1.000000      1.000000
50%        2.000000      1.000000
75%        2.000000      2.000000
max       10.000000      8.000000

### here above we can clearly see that mean of both Car and Bathroom columns are between 1 and 2, so we can replace/impute the Null values of Car and Bathroom either by 1 or 2to decide on either to impute with 1 or 2, we can see our data once which has highest values in the table is it 1 or 2we can use Count for this

### Method 1: -

In [63]:
df.groupby('Car').Car.count().sort_values(ascending = False)  # here it is clear that 2 is the highest assigned value

Car
2.0     5606
1.0     5515
0.0     1026
3.0      748
4.0      507
5.0       63
6.0       54
8.0        9
7.0        8
10.0       3
9.0        1
Name: Car, dtype: int64

### Method 2: -

In [65]:
df['Car'].astype('category').value_counts()

2.0     5606
1.0     5515
0.0     1026
3.0      748
4.0      507
5.0       63
6.0       54
8.0        9
7.0        8
10.0       3
9.0        1
Name: Car, dtype: int64

In [64]:
# |||ly we can Count for Bathroom column also

df.groupby('Bathroom').Bathroom.count().sort_values(ascending = False) # here it is clear that 1 is the highest assigned value

Bathroom
1.0    7517
2.0    4987
3.0     921
4.0     106
0.0      34
5.0      28
6.0       5
8.0       2
7.0       2
Name: Bathroom, dtype: int64

In [67]:
df.loc[:, 'Car'].fillna(2, inplace = True)

In [68]:
df.loc[:, 'Bathroom'].fillna(1, inplace = True)

In [69]:
round(df.isnull().sum().sort_values(ascending = False)/len(df) * 100, 2)

Propertycount    0.0
Regionname       0.0
Address          0.0
Rooms            0.0
Type             0.0
Price            0.0
Method           0.0
SellerG          0.0
Date             0.0
Distance         0.0
Postcode         0.0
Bedroom2         0.0
Bathroom         0.0
Car              0.0
Landsize         0.0
Lattitude        0.0
Longtitude       0.0
Suburb           0.0
dtype: float64

### Finally we do not have any Missing/Null values in our data

In [70]:
# Now lets check the final data that we have now after Cleansing all the Missing values

len(df)

13603

In [72]:
df.shape  # 23547 rows × 18 columns this was the shape of original data

(13603, 18)

### We can store original data into some other variable and our Cleans data into different variable to find out the final %ge of data remaining now for the Analysis

### ie; len(cleans_data)/len(og_data) * 100 = %ge of data remaining

-> 13603/23547 * 100 = 57.7695% of Data is remaining now.